<a href="https://colab.research.google.com/github/airpods69/DeepLearningGrind/blob/main/HeartDiseasePredictor.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Imports

In [ ]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, roc_curve, roc_auc_score, classification_report, f1_score
from sklearn.linear_model import LogisticRegression
import matplotlib.pyplot as plt
import seaborn as sns

## Kaggle related and Database download

In [ ]:
from google.colab import files
files.upload()

Saving kaggle.json to kaggle.json


{'kaggle.json': b'{"username":"airpods69","key":"6f30c3e40880f10e4bd9d237e14b09f8"}'}

In [ ]:
!mkdir ~/.kaggle && cp kaggle.json ~/.kaggle/ && chmod 600 ~/.kaggle/kaggle.json

In [ ]:
!kaggle datasets download -d johnsmith88/heart-disease-dataset

  0% 0.00/6.18k [00:00<?, ?B/s]
100% 6.18k/6.18k [00:00<00:00, 2.43MB/s]


In [ ]:
!unzip heart-disease-dataset.zip

Archive:  heart-disease-dataset.zip
  inflating: heart.csv               


## Analysis of DataSet

In [ ]:
data = pd.read_csv('heart.csv')

In [ ]:
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   52    1   0       125   212    0        1      168      0      1.0      2   
1   53    1   0       140   203    1        0      155      1      3.1      0   
2   70    1   0       145   174    0        1      125      1      2.6      0   
3   61    1   0       148   203    0        1      161      0      0.0      2   
4   62    0   0       138   294    1        1      106      0      1.9      1   

   ca  thal  target  
0   2     3       0  
1   0     3       0  
2   0     3       0  
3   1     3       0  
4   3     2       0

In [ ]:
data.shape

(1025, 14)

#### checking for empty coloumns

In [ ]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1025 entries, 0 to 1024
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1025 non-null   int64  
 1   sex       1025 non-null   int64  
 2   cp        1025 non-null   int64  
 3   trestbps  1025 non-null   int64  
 4   chol      1025 non-null   int64  
 5   fbs       1025 non-null   int64  
 6   restecg   1025 non-null   int64  
 7   thalach   1025 non-null   int64  
 8   exang     1025 non-null   int64  
 9   oldpeak   1025 non-null   float64
 10  slope     1025 non-null   int64  
 11  ca        1025 non-null   int64  
 12  thal      1025 non-null   int64  
 13  target    1025 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 112.2 KB


In [ ]:
data.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

As you can see no empty coloumns are present and the shape of data is 1025, 14
that is 1025 entries for 14 coloumns

## Model

a logistic regression model has been used from sklearn.linear_model

In [ ]:
x1 = data.drop(columns=['target'])
x = x1
y = data.target

In [ ]:
scaler = StandardScaler()
x_scaled = scaler.fit_transform(x)

In [ ]:
x_train, x_test, y_train, y_test = train_test_split(x_scaled,y,test_size = 0.25,random_state=100)

In [ ]:
log_reg = LogisticRegression()
log_reg.fit(x_train,y_train)

LogisticRegression()

In [ ]:
y_pred = log_reg.predict(x_test)

In [ ]:
accuracy_score(y_test,y_pred) 

0.8793774319066148

In [ ]:
confusion_matrix(y_test,y_pred)

array([[ 97,  22],
       [  9, 129]])

The Area Under the Curve (AUC) is the measure of the ability of a classifier to distinguish between classes and is used as a summary of the ROC curve. The higher the AUC, the better the performance of the model at distinguishing between the positive and negative classes.

In [ ]:
auc = roc_auc_score(y_test,y_pred)
auc

0.8749543295579102

F1 Score shows that whether there is overfitting in our model or not. The higher the value of F1 better the result which means there is very less overfitting.

In [ ]:
f1_score(y_test,y_pred)

0.8927335640138409

In [ ]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.92      0.82      0.86       119
           1       0.85      0.93      0.89       138

    accuracy                           0.88       257
   macro avg       0.88      0.87      0.88       257
weighted avg       0.88      0.88      0.88       257



In [50]:
k = np.array((5,0,0,100,248,0,0,122,0,1,1,0,2)).reshape(1, -1)
k_scaled = scaler.fit(x1)

In [53]:
ok = scaler.transform(k)

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but StandardScaler was fitted with feature names
  "X does not have valid feature names, but"


array([[-5.45157584, -1.51170646, -0.91575542, -1.80554022,  0.03878424,
        -0.41887792, -1.00404855, -1.17915797, -0.71228712, -0.06088839,
        -0.6241209 , -0.73197147, -0.52212231]])

In [52]:
log_reg.predict(ok)

array([1])